PayloadPkg/FirmwareUpdate/FirmwareUpdate.c

```diff
@@ -1163,6 +1165,73 @@ InitFirmwareUpdate (
   return EFI_SUCCESS;
 }
 
+
+/**
+  Try to recover a working boot partition.
+
+  @retval  EFI_SUCCESS          The operation completed successfully.
+  @retval  others               There is error happening.
+**/
+EFI_STATUS
+InitFirmwareRecovery (
```

Comment:
Check state machine within this function

Reply:
Add a new state for recovery. If we don't, on next boot, we attempt update (as opposed to recovery) since the failed boot counter is not persistent (but the states are).